lucene/core/src/java/org/apache/lucene/index/DocumentsWriter.java

```diff
@@ -273,33 +269,31 @@ synchronized Closeable lockAndAbortAll() throws IOException {
         pendingNumDocs.addAndGet(-ticket.getFlushedSegment().segmentInfo.info.maxDoc());
       }
     });
-    List<ThreadState> threadStates = new ArrayList<>();
+    List<DocumentsWriterPerThread> writer = new ArrayList<>();
     AtomicBoolean released = new AtomicBoolean(false);
     final Closeable release = () -> {
       if (released.compareAndSet(false, true)) { // only once
         if (infoStream.isEnabled("DW")) {
           infoStream.message("DW", "unlockAllAbortedThread");
         }
-        perThreadPool.unlockNewThreadStates();
-        for (ThreadState state : threadStates) {
+        perThreadPool.unlockNewWriters();
+        for (DocumentsWriterPerThread state : writer) {
```

Comment:
Isn't `writer` still an empty list here?  Oh I see, we are creating a lambda and only using it below, after we've added things to `writer`, OK.

Reply:
It might be less confusing if we reassign the `release` inside the `try` clause after acquiring DWPTs. Also nits: `writer` -> `writers` and `state` -> `writer`.

Reply:
unfortunately we can't assign this later otherwise the ref can't be final. I will leave a comment.